onnxruntime/core/providers/cuda/tensor/scatter_nd.cc

```diff
@@ -85,5 +101,88 @@ Status ScatterND::ComputeInternal(OpKernelContext* context) const {
   return Status::OK();
 }
 
+Status ScatterNDWithAtomicReduction::ComputeInternal(OpKernelContext* context) const {
+  const auto* input_tensor = context->Input<Tensor>(0);
+  const auto* indices_tensor = context->Input<Tensor>(1);
+  const auto* updates_tensor = context->Input<Tensor>(2);
+
+  const auto& input_shape = input_tensor->Shape();
+  const auto& indices_shape = indices_tensor->Shape();
+  const auto& updates_shape = updates_tensor->Shape();
+
+  // Validate input shapes
+  ORT_RETURN_IF_ERROR(onnxruntime::ScatterND::ValidateShapes(input_shape, indices_shape, updates_shape));
+
+  auto* output_tensor = context->Output(0, input_shape);
+
+  const void* input_data = input_tensor->DataRaw();
+  void* output_data = output_tensor->MutableDataRaw();
+
+  if (input_data != output_data) {
+    // TODO: Run benchmarks to determine if a dedicated kernel doing data copy will
+    // be faster than invoking cudaMemcpy ?
+    CUDA_RETURN_IF_ERROR(
+        cudaMemcpyAsync(output_data, input_data, input_tensor->SizeInBytes(),
+                        cudaMemcpyDeviceToDevice, Stream(context)));
+  }
+
+  // Bail out early
+  if (indices_shape.Size() == 0) {
+    return Status::OK();
+  }
+
+  auto last_index_dimension = indices_shape[indices_shape.NumDimensions() - 1];
+
+  // We need element counts for each dimension and the input dim value for each dimension
```

Comment:
I think, we can pass this by value within TArray<T> without additional calls to memcpy.

Reply:
I removed TensorShapeVector as it is difficult to know when it switches between stack and allocated tensors. I chose a more explicit way to avoid the memcpy.

Reply:
It is easy to know, just hover with the mouse, it would tell you `N=6` for int64_t, and you do not need to know. For shapes it would always be on the stack, for your case, it may be on the stack most of the time, and it gives toungilbe savings. It is also in the coding standard.

Reply:
I need to know because I cannot give the cuda kernel something allocated on cpu, only the stack or gpu is working. That's why I chose to create a struct with an explicit switch.

Reply:
I meant TArray. Never mind.